src/wallet.rs

```diff
@@ -0,0 +1,962 @@
+/// persisted in the wallet file.
+struct Addresses {
+    count: u64,
+}
+
+impl Default for Addresses {
+    fn default() -> Self {
+        Self { count: 1 }
+    }
+}
+
+impl Addresses {
+    pub(crate) fn from_bytes(buf: Vec<u8>) -> Result<Self, crate::Error> {
+        let mut u64_bytes = [0u8; 8];
+        u64_bytes.copy_from_slice(&buf);
+        Ok(Self {
+            count: u64::from_le_bytes(u64_bytes),
+        })
+    }
+
+    pub(crate) fn to_vec(&self) -> Vec<u8> {
+        self.count.to_le_bytes().to_vec()
+    }
+}
+
+/// Wrapper around `PathBuf` for wallet paths
+#[derive(PartialEq, Eq, Hash, Debug, Clone)]
+pub struct WalletPath(PathBuf);
+
+impl WalletPath {
+    /// Create a new wallet path from a directory and a name
+    pub fn new(dir: &Path, name: String) -> Self {
+        let mut pb = PathBuf::from(dir);
+        pb.push(name);
+        pb.set_extension("dat");
+        Self(pb)
+    }
+
+    /// Returns the filename of this path
+    pub fn name(&self) -> Option<String> {
+        // extract the name
+        let name = self.0.file_stem()?.to_str()?;
+        Some(String::from(name))
+    }
+
+    /// Returns current directory for this path
+    pub fn dir(&self) -> Option<PathBuf> {
+        let mut p = self.0.clone();
+        if p.pop() {
+            Some(p)
+        } else {
+            None
+        }
+    }
+
+    /// Sets the directory for the state cache
+    pub fn set_cache_dir(path: &Path) -> Result<(), Error> {
+        State::set_cache_dir(path.to_path_buf()).map_err(Error::from)
+    }
+
+    /// Wallet name defaults to user's username
+    pub fn default_name() -> String {
+        // get default user as default wallet name (remove whitespace)
+        let mut user: String = whoami::username();
+        user.retain(|c| !c.is_whitespace());
+        user
+    }
+
+    /// Wallet default directory defaults to user's home directory
+    pub fn default_dir() -> PathBuf {
+        let home = dirs::home_dir().expect("OS not supported");
+        Path::new(home.as_os_str()).join(DATA_DIR)
+    }
+
+    /// Checks if a wallet with this name already exists
+    pub fn exists(dir: &Path, name: &str) -> bool {
+        let mut pb = dir.to_path_buf();
+        pb.push(name);
+        pb.set_extension("dat");
+        pb.is_file()
+    }
+
+    /// Get full paths of all wallet files found in `dir`
+    pub fn wallets_in(dir: &Path) -> Result<Vec<WalletPath>, Error> {
+        let dir = fs::read_dir(dir)?;
+        let wallets = dir
+            .filter_map(|el| el.ok().map(|d| d.path()))
+            .filter(|path| path.is_file())
+            .filter(|path| match path.extension() {
+                Some(ext) => ext == "dat",
+                None => false,
+            })
+            .map(WalletPath)
+            .collect();
+
+        Ok(wallets)
+    }
+}
+
+impl Default for WalletPath {
+    fn default() -> Self {
+        let mut pb = Self::default_dir();
+        pb.push(Self::default_name());
```

Comment:
Shouldn't the `default_name` already include the `.dat` extension?

Reply:
Not really because we use the name for menus and input suggestions, while abstracting the `.dat` extension to the user!